File: packages/shared/types/statistics.ts
type VarLevel = 1 | 5 | 10;
type HorizonStats = { daily: number; monthly: number; annual: number };
type VaRByHorizon = { [H in 'daily' | 'monthly' | 'annual']: { [K in VarLevel]: number } };

export type Statistics = {
  cagr: number;
  mwrr: number;
  totalReturn: number;
  bestYear: number;
  worstYear: number;
  avgYear: number;
  avgAnnualReturn: number;
  avgMonthlyReturn: number;
  avgDailyReturn: number;

  stdev: number;
  stdevAnnual: number;
  stdevMonthly: number;
  stdevMonthlyRaw: number;
  stdevDaily: number;
  stdevDailyRaw: number;

  downsideDeviation: number;
  downsideDeviationDailyRaw: number;
  downsideDeviationMonthly: number;
  downsideDeviationMonthlyRaw: number;
  downsideDeviationAnnual: number;

  maxDrawdown: number;
  maxDrawdownDuration: number;
  avgDrawdown: number;
  ulcerIndex: number;
  drawdownRecoveryFactor: number;

  sharpe: number;
  sortino: number;
  calmar: number;
  ulcerPerformanceIndex: number;
  diversificationRatio: number;
  m2: number;
  treynor: number;

  alpha: number;
  beta: number;
  rSquared: number;
  benchmarkCorrelation: number;
  upsideCorrelation: number;
  downsideCorrelation: number;
  upsideBeta: number;
  downsideBeta: number;
  alphaDaily: number;
  alphaAnnualized: number;

  upsideCapture: number;
  downsideCapture: number;
  captureSpread: number;

  activeReturn: number;
  trackingError: number;
  informationRatio: number;

  var: VaRByHorizon;
  cvar: VaRByHorizon;

  varDaily5?: number;
  cvarDaily5?: number;
  skewness: HorizonStats;
  skewnessDaily?: number;
  excessKurtosis: HorizonStats;
  excessKurtosisDaily?: number;

  winRate: HorizonStats;
  pctPositiveDays: number;
  pctPositiveMonths: number;
  pctPositiveYears: number;

  maxDailyReturn: number;
  minDailyReturn: number;
  maxMonthlyReturn: number;
  minMonthlyReturn: number;
  maxAnnualReturn: number;
  minAnnualReturn: number;

  avgDailyGain: number;
  avgDailyLoss: number;
  gainLossRatioDaily: number;
  avgMonthlyGain: number;
  avgMonthlyLoss: number;
  gainLossRatioMonthly: number;
  avgAnnualGain: number;
  avgAnnualLoss: number;
  gainLossRatioAnnual: number;

  swr: number;
  pwr: number;
  swr10y: number;
  pwr10y: number;
  swr20y: number;
  pwr20y: number;
  swr30y: number;
  pwr30y: number;
  swr40y: number;
  pwr40y: number;
};

const ZERO_VAR: { [K in VarLevel]: number } = { 1: 0, 5: 0, 10: 0 };
const ZERO_SKEW: HorizonStats = { daily: 0, monthly: 0, annual: 0 };

type NonObjectKeys = {
  [K in keyof Statistics]-?: Statistics[K] extends number | undefined
    ? undefined extends Statistics[K]
      ? never
      : K
    : never;
}[keyof Statistics];

const ZERO_FIELDS: Record<NonObjectKeys, 0> = {
  cagr: 0,
  mwrr: 0,
  totalReturn: 0,
  bestYear: 0,
  worstYear: 0,
  avgYear: 0,
  avgAnnualReturn: 0,
  avgMonthlyReturn: 0,
  avgDailyReturn: 0,
  stdev: 0,
  stdevAnnual: 0,
  stdevMonthly: 0,
  stdevMonthlyRaw: 0,
  stdevDaily: 0,
  stdevDailyRaw: 0,
  downsideDeviation: 0,
  downsideDeviationDailyRaw: 0,
  downsideDeviationMonthly: 0,
  downsideDeviationMonthlyRaw: 0,
  downsideDeviationAnnual: 0,
  maxDrawdown: 0,
  maxDrawdownDuration: 0,
  avgDrawdown: 0,
  ulcerIndex: 0,
  drawdownRecoveryFactor: 0,
  sharpe: 0,
  sortino: 0,
  calmar: 0,
  ulcerPerformanceIndex: 0,
  diversificationRatio: 0,
  m2: 0,
  treynor: 0,
  alpha: 0,
  beta: 0,
  rSquared: 0,
  benchmarkCorrelation: 0,
  upsideCorrelation: 0,
  downsideCorrelation: 0,
  upsideBeta: 0,
  downsideBeta: 0,
  alphaDaily: 0,
  alphaAnnualized: 0,
  upsideCapture: 0,
  downsideCapture: 0,
  captureSpread: 0,
  activeReturn: 0,
  trackingError: 0,
  informationRatio: 0,
  pctPositiveDays: 0,
  pctPositiveMonths: 0,
  pctPositiveYears: 0,
  maxDailyReturn: 0,
  minDailyReturn: 0,
  maxMonthlyReturn: 0,
  minMonthlyReturn: 0,
  maxAnnualReturn: 0,
  minAnnualReturn: 0,
  avgDailyGain: 0,
  avgDailyLoss: 0,
  gainLossRatioDaily: 0,
  avgMonthlyGain: 0,
  avgMonthlyLoss: 0,
  gainLossRatioMonthly: 0,
  avgAnnualGain: 0,
  avgAnnualLoss: 0,
  gainLossRatioAnnual: 0,
  swr: 0,
  pwr: 0,
  swr10y: 0,
  pwr10y: 0,
  swr20y: 0,
  pwr20y: 0,
  swr30y: 0,
  pwr30y: 0,
  swr40y: 0,
  pwr40y: 0,
};

export function createEmptyStatistics(): Statistics {
  return {
    ...ZERO_FIELDS,
    var: { daily: { ...ZERO_VAR }, monthly: { ...ZERO_VAR }, annual: { ...ZERO_VAR } },
    cvar: { daily: { ...ZERO_VAR }, monthly: { ...ZERO_VAR }, annual: { ...ZERO_VAR } },
    skewness: { ...ZERO_SKEW },
    excessKurtosis: { ...ZERO_SKEW },
    winRate: { ...ZERO_SKEW },
  };
}

export function toStatsRecord(stats: Statistics): Record<string, number> {
  const record = stats as unknown as Record<string, number>;
  if (stats.var) {
    record.varDaily5 = stats.var.daily?.[5] ?? 0;
    record.cvarDaily5 = stats.cvar?.daily?.[5] ?? 0;
    record.varAnnual1 = stats.var.annual?.[1] ?? 0;
    record.varAnnual5 = stats.var.annual?.[5] ?? 0;
    record.cvarAnnual1 = stats.cvar?.annual?.[1] ?? 0;
    record.cvarAnnual5 = stats.cvar?.annual?.[5] ?? 0;
  }
  if (stats.skewness) record.skewnessDaily = stats.skewness.daily ?? 0;
  if (stats.excessKurtosis) record.excessKurtosisDaily = stats.excessKurtosis.daily ?? 0;
  record.negativeMonthsPct = 1 - stats.pctPositiveMonths;
  return record;
}

File: engine-go/internal/engine/statistics.go
package engine

import (
	"engine-go/internal/engineutil"
	"engine-go/internal/mathutil"
	"gonum.org/v1/gonum/stat"
	"math"
)

const (
	tradingDaysPerYear = engineutil.TradingDaysPerYear
	riskFreeRate       = engineutil.RiskFreeRate
)

type StatisticsRequest struct {
	Values                []float64  `json:"values"`
	Dates                 []string   `json:"dates"`
	StartingValue         float64    `json:"startingValue"`
	DailyReturns          []float64  `json:"dailyReturns"`
	AnnualReturnValues    []float64  `json:"annualReturnValues"`
	MonthlyReturnValues   []float64  `json:"monthlyReturnValues"`
	MwrrCashflows         []Cashflow `json:"mwrrCashflows"`
	BenchmarkDailyReturns []float64  `json:"benchmarkDailyReturns"`
	BenchmarkCagr         *float64   `json:"benchmarkCagr"`
}
type Cashflow struct {
	Value float64 `json:"value"`
	Time  float64 `json:"time"`
}

func calcRiskLevels(returns []float64, calc func([]float64, float64) float64) VaRLevels {
	return VaRLevels{One: calc(returns, 0.99), Five: calc(returns, 0.95), Ten: calc(returns, 0.90)}
}
func vaRByFrequency(returns [3][]float64, calc func([]float64, float64) float64) VaRByFrequency {
	return VaRByFrequency{Daily: calcRiskLevels(returns[0], calc), Monthly: calcRiskLevels(returns[1], calc), Annual: calcRiskLevels(returns[2], calc)}
}
func skewByFrequency(returns [3][]float64, calc func([]float64) float64) SkewnessByFrequency {
	return SkewnessByFrequency{Daily: calc(returns[0]), Monthly: calc(returns[1]), Annual: calc(returns[2])}
}
func CalculateStatisticsFromRequest(req StatisticsRequest) Statistics {
	if len(req.Values) < 2 {
		return Statistics{}
	}
	freqs := [3][]float64{req.DailyReturns, req.MonthlyReturnValues, req.AnnualReturnValues}
	finalValue := req.Values[len(req.Values)-1]
	years := float64(len(req.Dates)) / float64(tradingDaysPerYear)
	// 组合清零（finalValue<=0）时 CAGR 不可计算，置 0 而非 -1 哨兵，避免泄漏进 Sharpe/Calmar 等派生指标
	cagr := 0.0
	if finalValue > 0 {
		cagr = CalcCAGR(req.StartingValue, finalValue, years)
	}
	stdevDailyRaw := 0.0
	if len(req.DailyReturns) >= 2 {
		stdevDailyRaw = stat.StdDev(req.DailyReturns, nil)
	}
	stdevDaily := stdevDailyRaw * math.Sqrt(tradingDaysPerYear)
	rfDaily := RiskFreeDaily()
	dd := CalcMaxDrawdown(req.Values)
	ulcerIdx := CalcUlcerIndex(req.Values)
	sortino := CalcSortino(cagr, req.DailyReturns)
	sharpe := CalcSharpe(cagr, stdevDaily)
	mwrr := -1.0
	if finalValue > 0 {
		mwrr = CalcMWRR(append(append([]Cashflow{}, req.MwrrCashflows...), Cashflow{Value: finalValue, Time: years}))
	}
	var bm benchmarkMetrics
	if len(req.BenchmarkDailyReturns) >= 2 && req.BenchmarkCagr != nil {
		bm = computeBenchmarkMetrics(req.DailyReturns, req.BenchmarkDailyReturns, cagr, *req.BenchmarkCagr)
	}
	totalReturn := CalcTotalReturn(req.StartingValue, finalValue)
	pctPosDays := ratioPositive(req.DailyReturns)
	pctPosMonths := ratioPositive(req.MonthlyReturnValues)
	pctPosYears := ratioPositive(req.AnnualReturnValues)
	avgAnnual := stat.Mean(req.AnnualReturnValues, nil)
	avgMonthly := stat.Mean(req.MonthlyReturnValues, nil)
	avgDaily := stat.Mean(req.DailyReturns, nil)
	maxDailyRet := MaxValue(req.DailyReturns)
	minDailyRet := MinValue(req.DailyReturns)
	maxAnnualRet := MaxValue(req.AnnualReturnValues)
	minAnnualRet := MinValue(req.AnnualReturnValues)
	pwr := CalcPWR(req.AnnualReturnValues)
	pwrAll := CalcPWRAllYears(req.AnnualReturnValues)
	swr := 0.0
	for _, y := range []struct {
		threshold int
		val       float64
	}{{40, pwrAll.SWR40Y}, {30, pwrAll.SWR30Y}, {20, pwrAll.SWR20Y}, {10, pwrAll.SWR10Y}} {
		if len(req.AnnualReturnValues) >= y.threshold {
			swr = y.val
			break
		}
	}
	avgDailyGain, avgDailyLoss, gainLossRatioDaily := CalcAvgGainLoss(req.DailyReturns)
	avgMonthlyGain, avgMonthlyLoss, gainLossRatioMonthly := CalcAvgGainLoss(req.MonthlyReturnValues)
	avgAnnualGain, avgAnnualLoss, gainLossRatioAnnual := CalcAvgGainLoss(req.AnnualReturnValues)
	return Statistics{
		CAGR: cagr, MWRR: mwrr, Stdev: stdevDaily, Sharpe: sharpe, Sortino: sortino, MaxDrawdown: dd.MaxDrawdown, MaxDrawdownDuration: dd.MaxDrawdownDuration,
		BestYear: maxAnnualRet, WorstYear: minAnnualRet, AvgYear: avgAnnual, TotalReturn: totalReturn,
		MaxMonthlyReturn: MaxValue(req.MonthlyReturnValues), MinMonthlyReturn: MinValue(req.MonthlyReturnValues), AvgDrawdown: CalcAvgDrawdown(req.Values), UlcerIndex: ulcerIdx,
		Calmar: CalcCalmar(cagr, dd.MaxDrawdown), UlcerPerformanceIndex: CalcUPI(cagr, ulcerIdx), Beta: bm.Beta, Alpha: bm.Alpha, RSquared: bm.RSquared,
		TrackingError: bm.TrackingError, InformationRatio: bm.InformationRatio, UpsideCapture: bm.UpsideCapture, DownsideCapture: bm.DownsideCapture,
		MaxDailyReturn: maxDailyRet, MinDailyReturn: minDailyRet, PWR: pwr,
		Var: vaRByFrequency(freqs, CalcVaR), Cvar: vaRByFrequency(freqs, CalcCVaR),
		Skewness: skewByFrequency(freqs, CalcSkewness), ExcessKurtosis: skewByFrequency(freqs, CalcExcessKurtosis),
		WinRate: SkewnessByFrequency{Daily: pctPosDays, Monthly: pctPosMonths, Annual: pctPosYears}, PctPositiveDays: pctPosDays,
		AvgAnnualReturn: avgAnnual, AvgMonthlyReturn: avgMonthly, AvgDailyReturn: avgDaily,
		StdevAnnual: stat.StdDev(req.AnnualReturnValues, nil), StdevMonthly: stat.StdDev(req.MonthlyReturnValues, nil) * math.Sqrt(12), StdevMonthlyRaw: stat.StdDev(req.MonthlyReturnValues, nil),
		StdevDaily: stdevDaily, StdevDailyRaw: stdevDailyRaw,
		DownsideDeviation: mathutil.DownsideDeviation(req.DailyReturns, rfDaily) * math.Sqrt(tradingDaysPerYear), DownsideDeviationDailyRaw: mathutil.DownsideDeviation(req.DailyReturns, rfDaily),
		DownsideDeviationMonthly: mathutil.DownsideDeviation(req.MonthlyReturnValues, RiskFreeMonthly()) * math.Sqrt(12), DownsideDeviationMonthlyRaw: mathutil.DownsideDeviation(req.MonthlyReturnValues, RiskFreeMonthly()),
		DownsideDeviationAnnual: mathutil.DownsideDeviation(req.AnnualReturnValues, riskFreeRate), DrawdownRecoveryFactor: CalcDrawdownRecoveryFactor(totalReturn, dd.MaxDrawdown),
		M2: bm.M2, Treynor: bm.Treynor, BenchmarkCorrelation: bm.BenchmarkCorrelation, UpsideCorrelation: bm.UpsideCorrelation, DownsideCorrelation: bm.DownsideCorrelation,
		UpsideBeta: bm.UpsideBeta, DownsideBeta: bm.DownsideBeta, AlphaDaily: bm.AlphaDaily, AlphaAnnualized: bm.Alpha,
		CaptureSpread: bm.CaptureSpread,
		ActiveReturn:  bm.ActiveReturn, PctPositiveMonths: pctPosMonths, PctPositiveYears: pctPosYears, MaxAnnualReturn: maxAnnualRet, MinAnnualReturn: minAnnualRet,
		AvgDailyGain: avgDailyGain, AvgDailyLoss: avgDailyLoss, GainLossRatioDaily: gainLossRatioDaily,
		AvgMonthlyGain: avgMonthlyGain, AvgMonthlyLoss: avgMonthlyLoss, GainLossRatioMonthly: gainLossRatioMonthly,
		AvgAnnualGain: avgAnnualGain, AvgAnnualLoss: avgAnnualLoss, GainLossRatioAnnual: gainLossRatioAnnual,
		SWR: swr, SWR10Y: pwrAll.SWR10Y, PWR10Y: pwrAll.PWR10Y, SWR20Y: pwrAll.SWR20Y, PWR20Y: pwrAll.PWR20Y, SWR30Y: pwrAll.SWR30Y, PWR30Y: pwrAll.PWR30Y, SWR40Y: pwrAll.SWR40Y, PWR40Y: pwrAll.PWR40Y,
	}
}
